### ISSUE TITLE
Proposal: IncrementalHashAlgorithm

### ISSUE THREAD

**Issue by Metalnem** [ektrah/nsec#4]
Bytes used as an input to the HashAlgorithm are often not in a single span. For that purpose the .NET standard library has the IncrementalHash class. I propose implementing something similar in NSec. Base class would in that case probably look something like this:

```csharp
public sealed class IncrementalHashAlgorithm : Algorithm, IDisposable
{
	public void AppendData(ReadOnlySpan<byte> data);
 	public bool TryGetHashAndReset(Span<byte> destination, out int bytesWritten);
}
```

(This is the way IncrementalHash is implemented in .NET Core 2.1, see dotnet/corefx#22614.)

In case that you approve this addition to the API, I would be happy to submit the pull request.

**Comment by ektrah** [comment]
That sounds like a useful addition (and is actually on my to-do list). I'd like to iterate a bit on the API before implementation starts.

First, zero-allocation. Incremental hashing might be used in performance-critical code and shouldn't require the allocation of a class instance. A struct would be the obvious alternative, but mutable structs are evil&trade;.

Second, there are some interesting additions to the System.Buffers namespace in the System.Memory package. It seems one option could be letting the hash algorithm pull from a sequence of spans rather than pushing several spans to it.

**Comment by Metalnem** [comment]
I don't think you can avoid the allocation in this case. You have to keep the state of the hash function across the update calls, but you can't stackalloc it like you did in the HashAlgorithm.

Sequence of spans could work, but that would be slightly weird API for the users (every library I've seen so far has some version of Update/Finish functions).

**Comment by ektrah** [comment]
Added two more items.

I was thinking if something like this could work:

```csharp
Span<IncrementalHashState> state = stackalloc IncrementalHashState[1];
algorithm.Init(state);
algorithm.Update(state, span1);
algorithm.Update(state, span2);
algorithm.Update(state, span3);
algorithm.Final(state, hash);
```

Or maybe a mutable struct wouldn't be that evil at all:

```csharp
algorithm.Init(out IncrementalHashState state);
algorithm.Update(ref state, span1);
algorithm.Update(ref state, span2);
algorithm.Update(ref state, span3);
algorithm.Final(ref state, hash);
```

(where `IncrementalHashState` in both cases is a struct without public members).

I agree that a sequence accepting API would be somewhat weird for this use case (though maybe a nice addition to the single-span and update/finish APIs).

**Comment by Metalnem** [comment]
That would work, but it breaks encapsulation, and it's very ugly approach, especially compared to the rest of the library, which is super simple and elegant.

What about just embedding the state (which would be a struct) in the algorithm object? You have to allocate the algorithm object anyway, and it doesn't really matter if its size is 8 bytes or 208 bytes. Something like this:

```csharp
[StructLayout(LayoutKind.Sequential, Size = 104)]
public struct Sha256State
{
}

public class Sha256Algorithm
{
	private readonly Sha256State state;
}
```

**Comment by ektrah** [comment]
You've discovered why incremental hashing isn't already part of the library :-)

Embedding the state is somewhat ugly, since all algorithm classes are stateless right now - and therefore thread-safe. (You're not expected to allocate a new algorithm instance for each operation.) And it doesn't satisfy the alignment requirement of Blake2b (which my proposal above doesn't as well, sadly).

The alignment requirements basically means that the state must be pinned the whole time or copied on each update.

**Comment by ektrah** [comment]
Proposal:

```csharp
var hash = new byte[sha256.DefaultHashSize];
IncrementalHash.Initialize(sha256, hash.Length, out var state);
IncrementalHash.Update(ref state, span1);
IncrementalHash.Update(ref state, span2);
IncrementalHash.Update(ref state, span3);
IncrementalHash.Finalize(ref state, hash);
```

```csharp
namespace NSec.Cryptography
{
    public struct IncrementalHash
    {
        public static void Initialize(HashAlgorithm algorithm, int hashSize, out IncrementalHash state);
        public static void Update(ref IncrementalHash state, ReadOnlySpan<byte> data);
        public static byte[] Finalize(ref IncrementalHash data);
        public static void Finalize(ref IncrementalHash state, Span<byte> hash);
        public static void FinalizeAndVerify(ref Incremental state, ReadOnlySpan<byte> hash);
        public static bool FinalizeAndTryVerify(ref Incremental state, ReadOnlySpan<byte> hash);
    }

    public struct IncrementalMac
    {
        public static void Initialize(MacAlgorithm algorithm, Key key, int macSize, out IncrementalMac state);
        public static void Update(ref IncrementalMac state, ReadOnlySpan<byte> data);
        public static byte[] Finalize(ref IncrementalMac data);
        public static void Finalize(ref IncrementalMac data, Span<byte> mac);
        public static void FinalizeAndVerify(ref Incremental state, ReadOnlySpan<byte> mac);
        public static bool FinalizeAndTryVerify(ref Incremental state, ReadOnlySpan<byte> mac);
    }
}
```

**Comment by ektrah** [comment]
@Metalnem Do you have an opinion about the above API shape? It's not an evil mutable struct, doesn't break encapsulation, and looks relatively simple to me.

**Comment by Metalnem** [comment]
Most of what I've said in the previous comment still holds (the state remains exposed instead of being hidden inside the object instance).

Besides that, this API looks more low-level than the rest of the library (it's basically the direct translation of the C API).

In my opinion, it all boils down to this: in this case you can either have a zero-allocation API, or a standard library quality API. You are in favor of the former, I'm in favor of the latter :)

**Comment by Metalnem** [comment]
Not sure if the belongs to this issue, but I don’t think that you need all the Verify functions to be instance methods. I believe they are all doing the same thing - caling some constant-time memory compare function? If that’s the case, maybe they could be just static methods in the HashAlgorithm/MacAlgorithm class?

**Comment by ektrah** [comment]
@Metalnem Thanks for taking a look.

I'm trying to argue that the difference isn't actually that big :smile:. In fact, the only difference is that you write `StaticMethod(ref state)` instead of `state.InstanceMethod()`. The state is hidden in the instance either way; you don't get more access to the internal state than with the other.

You get a bit more control of how the state is passed around (by-reference or by-value) and zero-allocation with structs, which I think is desirable.

---

The Verify methods indeed just provide a hash/MAC calculation with a constant-time memory compare in one step. They essentially exist only to make constant-time compare easier to use than the built-in SequenceEqual method. Of course, I could simplify the API a bit by letting users do the verification in two steps, but I think the convenience methods are a small win here.

The one step verification methods need know which algorithm you want to use. If they were static, you'd need to pass the algorithm to them as a parameter. Here, I think `algorithm.InstanceMethod()` is more discoverable than `StaticMethod(algorithm)`, because the primary Hash/Mac methods are (and must be) instance methods.

Feedback like this is *very* welcome, btw, while there's still time for breaking changes until System.Memory gets stable. I see you're working on a Noise implementation; NSec is exactly intended for use cases like that.  Do you see any other issues that would prevent NSec from being used to (theoretically speaking) replace the crypto algorithm classes in your code?

**Comment by Metalnem** [comment]
The difference may look small, but if you are aiming for the wide adoption of the library, you should probably provide the API that feels natural. I will use this library in any case, but other users may end up confused.

Do you really need to know the algorithm for the verification? You only have to compare the two byte spans, irrespective of the algorithm that produced them. Anyway, this is more of a nitpick than a real issue (it's just that I always prefer the smallest API surface possible).

I was waiting for the NSec to be released to start working on my Noise implementation, but finally gave up because I didn't want to depend on the release of .NET Core 2.1 (you released the first preview version after I started my implementation). I'm currently directly using [libsodium](https://www.nuget.org/packages/libsodium/), but I'll experiment with replacing my interop code with NSec some time this week. I'll give you more feedback when that happens.
